src/box/lua/config/utils/odict.lua

```diff
@@ -0,0 +1,209 @@
+
+local REINDEX_THRESHOLD = 1000
+
+local registry = setmetatable({}, {__mode = 'k'})
+
+-- Generator function for odict.pairs().
+--
+-- Returns (key, value).
+local function gen(od, prev_key)
+    local ctx = registry[od]
+
+    -- The previous key is nil only on the first call of the
+    -- generator function.
+    local id = prev_key == nil and 0 or ctx.key2id[prev_key]
+    -- NB: This assert doesn't catch all the kinds of changes
+    -- during an iteration. It rather verifies a precondition
+    -- for the following cycle.
+    assert(id ~= nil, 'ordered dictionary is changed during iteration')
+
+    while id <= ctx.max_id do
+        id = id + 1
+        local key = ctx.id2key[id]
+        -- id2key may have holes for IDs <= max_id in place of
+        -- deleted items.
+        --
+        -- id2key may contain a stalled entry, because __newindex
+        -- is not called on assignment of an existing field,
+        -- including assignment to nil.
+        if key ~= nil and od[key] ~= nil then
+            return key, od[key]
+        end
+    end
+end
+
+-- Returns a Lua iterator triplet (gen, param, state).
+--
+-- Usage:
+--
+--  | for k, v in odict.pairs(od) do
+--  |     <...>
+--  | end
+--
+-- The iterator stability guarantees are the same as for usual
+-- pairs().
+--
+-- Quote from the Lua 5.1 reference manual:
+--
+-- > The behavior of `next` is undefined if, during the traversal,
+-- > you assign any value to a non-existent field in the table.
+-- > You may however modify existing fields. In particular, you
+-- > may clear existing fields.
+local function _pairs(od)
+    return gen, od, nil
+end
+
+-- Delete the key from the id<->key mappings.
+local function trash(ctx, key)
+    local id = ctx.key2id[key]
+    if id ~= nil then
+        ctx.key2id[key] = nil
+        ctx.id2key[id] = nil
+
+        -- No need to grow the IDs if the same key is assigned and
+        -- deleted in a loop.
+        --
+        -- for <...> do
+        --     od.x = 'x'
+        --     od.x = nil
+        -- end
+        if id == ctx.max_id then
+            ctx.max_id = ctx.max_id - 1
+        end
+    end
+end
+
+-- Track the new key in the id<->key mappings.
+local function track(ctx, key)
+    local id = ctx.max_id + 1
+    ctx.id2key[id] = key
+    ctx.key2id[key] = id
+    ctx.max_id = id
+end
+
+-- Renew the id<->key mappings to eliminate holes.
+local function reindex(od, ctx)
+    -- This cycle reassigns IDs in the following way.
+    --
+    --    1     2     3     4
+    -- | foo | nil | bar | baz |
+    --          ^    | ^    |
+    --          +----+ +----+
+    local old_max_id = ctx.max_id
+    ctx.max_id = 0
+    for id = 1, old_max_id do
+        local key = ctx.id2key[id]
+        -- Drop the given key-id pair from the key<->id mappings.
+        if key ~= nil then
+            trash(ctx, key)
+        end
+        -- Add the new key-id pair into the key<->id mappings.
+        if key ~= nil and od[key] ~= nil then
```

Comment:
Nit:
I may be mistaken, but it seems like we don't need `key ~= nil` check, do we?

Reply:
Yep. We can use `nil` as a table index key and will always get `nil` as a result (but we can't assign a `nil` key). However, I like this condition to be written explicitly, because:

1. There is the `key ~= nil` condition above and it is easier to spot that both `trash()` and `track()` are skipped in case of `key == nil`.
2. Using `nil` as an index key for a table is a trick (not known by any Lua user) and it anyway should be commented. An explicit check looks better for me than an implicit one + comment.

Reply:
If you prefer so, sure.